accepted/2025/supporting-bearer-auth.md

```diff
@@ -0,0 +1,72 @@
+When credential providers return a bearer token, the request will be retried by adding the bearer token to the authentication header.
+When we make a request with an expired token, the server should return a 401 Unauthorized respond, causing a new request for credentials.
+
+### Technical explanation
+
+[HttpSourceAuthenticationHandler.SendAsync](https://github.com/NuGet/NuGet.Client/blob/313eecd3af442ee2eeed2e6decf310858934ab21/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceAuthenticationHandler.cs?plain=1#L68) will be updated to check if the previous response contains a WWW-Authenticate: Bearer header.
+If it does, then before sending the request the method will attempt to retrieve the token from the credentials provided and pass them in the authentication header if it’s available.
+
+[GetAuthenticationCredentialsRequest](https://github.com/NuGet/NuGet.Client/blob/ab6d3b886b32e9b8f68a7f1b299c43b0c72487dc/src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsRequest.cs?plain=1#L12C25-L12C60) will be updated to include a new field, `WwwAuthenticateHeader`, that will tell the credential providers which auth headers can be used.
+This can be used by CredentialProviders to know if they’re communicating with a version of NuGet that supports Bearer tokens.
+If `WwwAuthenticateHeader` is null, then the version of nuget they are communicating with does not support bearer tokens. Credential providers that want to use bearer auth will return a [GetAuthenticationCredentialsResponse](https://github.com/NuGet/NuGet.Client/blob/ab6d3b886b32e9b8f68a7f1b299c43b0c72487dc/src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs?plain=1#L14C1-L14C61) that has “bearer” in the AuthenticationTypes list.
+
+When deciding which protocol to use for authentication we will use a combination of the headers provided and the results from ICredentials.GetCredential.
+If the header indicates the server accepts basic, and GetCredential returns a result for basic, we will use that for the request.
+Similarly, if the header accepts bearer, and GetCredentials returns a result for bearer, we will use that.
+
+To maintain the “PreAuthenticate” functionality, a cache can be used to keep track of URLs and the credential previously used to authenticate against it.
+
+## Drawbacks
+
+Even though we are moving away from sending the bearer token as a password during the request, the ICredential.GetCredential explicitly returns a NetworkCredential, which requires a username and password.
+
+## Rationale and alternatives
+
+### Alt - Use basic Authentication
+
+This option would have NuGet.org implement the credential provider like Azure Artifacts does for authentication.
+The client wouldn’t need to be updated, and the API Key would be sent over to the server using basic authentication.
+The username field would be filled with a placeholder.
```

Comment:
NuGet client wouldn't fill the username with a placeholder, the cred provider would need to.